### ISSUE TITLE
List is pasted from OneNote in a reversed order

### ISSUE THREAD

**Issue by dryzhkov** [ckeditor/ckeditor-dev#796]
## Are you reporting a feature request or a bug?

Bug

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

*Put all reference links here…*

## Provide detailed reproduction steps (if any)

1. Create a bulleted list in OneNote. Ex:

- One
- Two
- Three

2. Select all content with ctrl-a
3. Paste into the latest CKEditor with ctrl-v (Ex: https://ckeditor.com/features)

### Expected result

Order should be preserved (One -> Two -> Three)

### Actual result

Order is reversed (Three -> Two -> One)

## Other details
Note: this seems to happen only if the outer <li> is selected.

* Browser: Chrome
* OS: Mac , Windows
* CKEditor version: 4.7.1+ (Seems to be working in older versions)
* Installed CKEditor plugins: …

**Comment by beatadelura** [comment]
@dryzhkov Thank you for creating this issue. Unfortunately, we don't support `copy/paste` functionality from OneNote, so I have to close this.

**Comment by WrathOfZombies** [comment]
Thanks for the response @beatadelura. Unfortunately this is critical scenario for us and upon debugging further we observed that the reversal happens inside of https://github.com/ckeditor/ckeditor-dev/blob/master/core/htmlparser/fragment.js#L476 which is invoked from https://github.com/ckeditor/ckeditor-dev/blob/master/plugins/pastefromword/filter/default.js#L2167.

What I am unable to understand is why do we do a reversal of the list and why is it necessary? This is a scenario that's only observed when the HTML dom that's provided to CKEditor looks like follows:

```html
<ul>
    <!-- Seems that CKEditor creates a fake li here when it doesn't find any children -->
    <ul>
        <li></li>
        <li></li>
        <li></li>
    </ul>
</ul>
```

**Comment by Stalq** [comment]
Hello @WrathOfZombies and @dryzhkov. Thank you very much for your message.
This is Rafal - Head of Sales in CKSource. Most likely the problem with copying/pasting from OneNote is caused by they way how the markup is provided by this application to the browser clipboard. If it happens that it is not consistent with what CKEditor expects then we would need to make a research on this topic. Nevertheless, we are sure that this is not a bug, but rather new feature request. I will reach out to you @WrathOfZombies  and @dryzhkov via our support channel. Thanks.

**Comment by mlewand** [comment]
We don't suport OneNote but let's see how the markup is actually wrong, then based on this we can say what we want to do with it.

Reopening the issue.

**Comment by jswiderski** [comment]
## Provide detailed reproduction steps (if any)

1. Create a 3 element list in OneNote page
2. While cursor blinks in second or third element use Ctrl + A twice to select the entire list (The first Ctrl + A selects single list item in OneNote)
3. Press Ctrl + C
4. Paste into CKEditor with Ctrl + V

### Expected result

List gets pasted as it was created in OneNote

### Actual result

List gets pasted with elements in reverse order

## Other details

* Browser: Chrome, Firefox
* OS: Win
* CKEditor version: 4.7.0+ (I could not reproduce this issue in CKEditor 4.6.2 or below)
* Installed CKEditor plugins: N/A

------

I have tested this issue in OneNote 2010 and 2016.

The list pasted from OneNote looks like below:

```
<ul style="margin-left:.2847in;direction:ltr;unicode-bidi:embed;margin-top:
 0in;margin-bottom:0in">
 <ul type="disc" style="margin-left:.375in;direction:ltr;unicode-bidi:embed;
  margin-top:0in;margin-bottom:0in">
  <li style="margin-top:0;margin-bottom:0;vertical-align:middle"><span style="font-family:Calibri;font-size:11.0pt">123</span></li>
  <li style="margin-top:0;margin-bottom:0;vertical-align:middle"><span style="font-family:Calibri;font-size:11.0pt">456</span></li>
  <li style="margin-top:0;margin-bottom:0;vertical-align:middle"><span style="font-family:Calibri;font-size:11.0pt">789</span></li>
 </ul>
</ul>
```

As you can see there is double UL element what results in the editor as list inside the list element. This result can still be observed in IE11 and Edge while it has gone in Chrome in version 4.7.0. Maybe it has something to do with reversing the order. In Firefox till version 4.6.2, list was pasted as image inside paragraph. Starting from version 4.7.0 you get clean HTML but with reversed list items.

**Comment by jswiderski** [comment]
[OneNoteIssue.zip](https://github.com/ckeditor/ckeditor-dev/files/1270217/OneNoteIssue.zip)

Sample screen-cast

**Comment by mlewand** [comment]
Actually looks that it's a bug on our side @msamsel will elaborate more on that.

**Comment by msamsel** [comment]
Hi,

I hope I find the problem.
There is part of our code which process indent list. In case when unnecessary indention are reduced, children nodes are added to the list always at index `0`. So it results with opposite order of list items.
https://github.com/ckeditor/ckeditor-dev/blob/da389f9f08a8df4562ce14ae02ad215dd510c454/plugins/pastefromword/filter/default.js#L2166-L2168
I'll prepare pull request which will fix this behaviour.

**Comment by mlewand** [comment]
Fixed by #863.

**Comment by WrathOfZombies** [comment]
@mlewand Thanks. Testing this out internally and I'll get back in case we have further questions.
